### PR TITLE
feat: Add negative years support 

### PR DESCRIPTION
**Pull request by klm-turing**
Add ability to parse negative years

This fix [2636](https://github.com/iamkun/dayjs/issues/2636) issue

### CODE REVIEW

**Comment by willismonroe** [comment]
This looks good, I added a test (in manipulate.test.js):
```
it('Add and subtract with negative years', () => {
  expect(dayjs('-2006').add(1, 'y')).toEqual(dayjs('-2005'))
  expect(dayjs('-2006').subtract(1, 'y')).toEqual(dayjs('-2007'))
})
```
Just to make sure that adding/subtracting were working well. I tried to commit to your fork, but it didn't work.

**Comment by klm-turing** [comment]
I added those test for you
Screenshot from 2024-04-26 17-41-41

**Comment by willismonroe** [comment]
That looks good!

**Comment by klm-turing** [comment]
Hey Dear @willismonroe It's now a plugin to be extended as @iamkun suggested

File: test/plugin/negativeYear.test.js [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+describe('negativeYear', () => {
+  it('parses negative years', () => {
+    expect(dayjs('-2020-01-01').year()).toBe(-2020)
+    const date = '-2021/01/03'
+    const date2 = '01/03/-2021'
+    const date3 = '01-03--2021'
+    const d = date.match(REGEX_PARSE)
+    expect(dayjs(date).format('YYYY-MM-DD')).toBe(`-${moment(date).format('YYYY-MM-DD')}`)
+    expect(dayjs(date).format()).toBe(`-${moment(date).format()}`)
```
Comment (iamkun) [inline test/plugin/negativeYear.test.js:28]: we could remove moment in this test case, as far as i know, moment does not support nagative year

File: src/plugin/negativeYear/index.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+  const parseDate = (cfg) => {
+    const { date } = cfg
+    if (!date) return new Date()
+    const newDate = new Date(date)
+    const fullYear = newDate.getFullYear()
+    if (typeof date === 'string' && date.indexOf(`-${fullYear}`) !== -1) {
+      return dayjs(newDate).subtract(fullYear * 2, 'year').toDate()
+    }
+    return newDate
```
Comment (iamkun) [inline src/plugin/negativeYear/index.js:12]: we can return the passed in the original cfg.date directly, without dealing with logic like !date

Comment (klm-turing) [reply src/plugin/negativeYear/index.js:12]: @iamkun Got you

File: test/plugin/negativeYear.test.js (hunk 1 above)
Comment (klm-turing) [reply test/plugin/negativeYear.test.js:28]: @iamkun Sure

**Comment by klm-turing** [comment]
@iamkun Changes are done.

File: src/plugin/negativeYear/index.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+  const proto = c.prototype
+
+  const parseDate = (cfg) => {
+    const { date } = cfg
+    const newDate = new Date(date)
+    const fullYear = newDate.getFullYear()
+    if (typeof date === 'string' && date.indexOf(`-${fullYear}`) !== -1) {
+      return dayjs(newDate).subtract(fullYear * 2, 'year').toDate()
+    }
+    return newDate
```
Comment (iamkun) [inline src/plugin/negativeYear/index.js:11]: maybe we should return date not newDate, cause `cfg.date` is the untouched original one, just leave it to the current main parse logic

Comment (klm-turing) [reply src/plugin/negativeYear/index.js]: @iamkun Updated

**Comment by iamkun** [comment]
THX

**Pull request merged**